packages/theia-extension/cdtcloud/src/node/compilation-service.ts

```diff
@@ -0,0 +1,59 @@
+import { injectable } from "inversify";
+import { CompilationService } from "../common/protocol";
+import { readdir } from 'fs/promises';
+import { createReadStream } from 'fs';
+import { RPCClient } from "./rpc-client";
+import 'reflect-metadata';
+import FormData from 'form-data';
+import { join } from 'path';
+import got from "got";
+
+@injectable()
+export class CompilationServiceImpl implements CompilationService {
+
+  binaryFile: string;
+  artifactUri: string;
+
+  async compile(fqbn: string, id: string, sketchPath: string): Promise<void> {
+    const client = new RPCClient()
+    await client.init()
+    await client.createInstance()
+    await client.initInstance()
+
+    const buildPath = await client.getBuildPath(fqbn, sketchPath)
+    const files = await readdir(buildPath)
+
+    files.forEach((file) => {
+      if (file.endsWith('.bin')) {
+        this.binaryFile = join(buildPath, file)
+      }
+    })
```

Comment:
1) You can break once the right file was found
2) Compiling can give you different files with the same extension. Make sure you select the right one (not ending in with_bootloader.{bin|hex})
3) Files extensions depend on the arduino core type

Something like this **Not tested**
```suggestion
// Method to determine required extension
const getRequiredExtension = async (fqbn: string): Promise<string> => {
  if (fqbn.startsWith('arduino')) {
    const segments = fqbn.split(':')
    switch (segments[1]) {
      case 'avr':
        return '.hex'
      case 'sam':
        return '.bin'
      case 'mbed_edge':
      case 'mbed_nano':
      case 'mbed_nicla':
      case 'mbed_portenta':
      case 'mbed_rp2040':
      case 'megaavr':
      case 'nrf52':
      case 'samd':
      case 'mbed':
        throw new Error('No extension specified')
      default:
        throw new Error('Unknown arduino board core')
    }
  }
  throw new Error('Unknown device')
}

const buildPath = await client.getBuildPath(fqbn, sketchPath)
const files = await readdir(buildPath)
const ext = await getRequiredExtension(fqbn)

files.forEach((file) => {
  if (Path.extname(file) === ext && !file.endsWith(`with_bootloader${ext}`)) {
    this.binaryFile = join(buildPath, file)
    break
  }
}

```